component/phlare/write/write.go

```diff
@@ -0,0 +1,263 @@
+package write
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/bufbuild/connect-go"
+	"github.com/go-kit/log/level"
+	"github.com/oklog/run"
+	commonconfig "github.com/prometheus/common/config"
+	"github.com/prometheus/prometheus/model/labels"
+	"go.uber.org/multierr"
+
+	"github.com/grafana/agent/component"
+	"github.com/grafana/agent/component/common/config"
+	pushv1 "github.com/grafana/agent/component/phlare/push/v1"
+	pushv1connect "github.com/grafana/agent/component/phlare/push/v1/pushv1connect"
+	"github.com/grafana/agent/component/pprof"
+	"github.com/grafana/agent/pkg/build"
+)
+
+var (
+	userAgent        = fmt.Sprintf("GrafanaAgent/%s", build.Version)
+	DefaultArguments = func() Arguments {
+		return Arguments{}
+	}
+	DefaultEndpointOptions = func() EndpointOptions {
+		return EndpointOptions{
+			RemoteTimeout: 30 * time.Second,
+		}
+	}
+	_ component.Component = (*Component)(nil)
+)
+
+func init() {
+	component.Register(component.Registration{
+		Name:    "phlare.write",
+		Args:    Arguments{},
+		Exports: Exports{},
+		Build: func(o component.Options, c component.Arguments) (component.Component, error) {
+			return NewComponent(o, c.(Arguments))
+		},
+	})
+}
+
+// Arguments represents the input state of the phlare.write
+// component.
+type Arguments struct {
+	ExternalLabels map[string]string  `river:"external_labels,attr,optional"`
+	Endpoints      []*EndpointOptions `river:"endpoint,block,optional"`
+}
+
+// UnmarshalRiver implements river.Unmarshaler.
+func (rc *Arguments) UnmarshalRiver(f func(interface{}) error) error {
+	*rc = DefaultArguments()
+
+	type config Arguments
+	return f((*config)(rc))
+}
+
+// EndpointOptions describes an individual location for where profiles
+// should be delivered to using the Phlare push API.
+type EndpointOptions struct {
+	Name             string                   `river:"name,attr,optional"`
+	URL              string                   `river:"url,attr"`
+	RemoteTimeout    time.Duration            `river:"remote_timeout,attr,optional"`
+	Headers          map[string]string        `river:"headers,attr,optional"`
+	HTTPClientConfig *config.HTTPClientConfig `river:"http_client_config,block,optional"`
+}
+
+// UnmarshalRiver implements river.Unmarshaler.
+func (r *EndpointOptions) UnmarshalRiver(f func(v interface{}) error) error {
+	*r = DefaultEndpointOptions()
+
+	type arguments EndpointOptions
+	return f((*arguments)(r))
+}
+
+// Component is the prometheus.remote_write component.
+type Component struct {
+	opts component.Options
+	cfg  Arguments
```

Comment:
I think in most our components we blindly add a mutex here, but just making sure we don't need it here, right? Every time new arguments are passed we only use them in one goroutine to recreate the receiver.

Reply:
Checking .....

Reply:
Yeah doesn't seems necessary to me.